added support for paging via API methods

pionl opened the pull request:
Hi,
I’ve noticed that paging doesn’t work on API methods (not implemented). I’ve added support to enable paging on the API calls (thread, categories, posts). Only category list is not perfect - there is a filter which removes private categories after the query is done. The paginate will return data, but the total count is not accurate (based on the private settings).

We can disable paging for categories by setting 0.

To enable paging set `true` in the config: `forum.preferences.pagination.enabled`

Commit:
added support for paging via API methods

Commit:
added missing requests scopes for posts API

pionl commented:
hi @Riari , I've also added a support for setting custom models. Its prepared in new branch and goes from this branch. Will make PR when merged? https://github.com/pionl/laravel-forum/tree/api-custom-models

Riari commented:
Hey,

I've been very busy lately so haven't been able to look at this properly yet, but I like the principle of it. I'll try to get to it this weekend and merge if it all looks good.

Thanks!

pionl commented:
@Riari no hurry, I've got a lot work too. I'm using my own branch at this moment. Thanks for the package works nicely! At this moment I think I've changed all that I need to extends/override 👍 

Have a great day!

Commit:
when using paginating, the posts must be returned as a tree (to support reply)

this helps to support paging as much as possible, there could be
children in paginate too (could be overkill)

Commit:
added support to pass the request scopes into children relation

added scopeWithRequestScopes
updated controllers to pass request object

Riari commented:
Sorry for the delay. I see there have been some more commits; can you split them into separate branches and do a new PR? I'd like to see the pagination changes in isolation, but feel free to do a PR for other stuff too. Thanks!

pionl commented:
@Riari Hi, the new commits depends on the paginate refactor and improvs the paginate to work more logically.

The `926b1b43fe47eee837ba0a5ac7a1b9f27418cfbd` commit fixes add support for children's - it's impossible to return correct posts when paginating is in order and passing the already implemented scopeWithRequest

I've got more tweaks in other branches, but need this PR to be merged, should I post.

I can then make 2 branches:
paginate (5379af2 and   a2be250) and then improvment branch (return post tree when paginate) (aa47334 and 926b1b4)

Riari commented:
My mistake, I'll re-open this - but on closer inspection I do have some doubts about how it's implemented. Give me some time to play with these changes and I'll get back to you. 👍

pionl commented:
@Riari ok np, a lot of calls were same so I used base calls that uses config. I't can be changed to use different approach

Riari commented:
Okay, I've had a chance to look at this more closely. I appreciate the time put into this so far, but I don't want to accept the current changes. Reasons are as follows:

1. I don't think there's any real justification for enabling pagination on categories. I know the models and config are already set up for it, but considering this PR made me realise it's kind of pointless. I can't imagine any forum being so large as to really warrant paginating categories, even in an admin UI (and from memory, I don't believe I've seen any of the major forum platforms do it). So my inclination is to keep it simple and remove pagination from categories altogether.
2. I feel like the implementation could be a lot simpler - with the above in mind, it probably could be. In theory, all that needs to happen is for a `page` querystring param to be passed through and used accordingly in the index methods for threads and posts.

I'm happy to look at this stuff more closely and try a few ideas out, but I don't know when that will be. If you want to submit a new PR (or update this one) with this feedback in mind, feel free. Just let me know. Thanks!

pionl commented:
Hi @Riari, if I recall correctly, there was already settings for it but without any logic for it. Second point - ok, I can make this kind of change. Good point.

Martin.

Riari commented:
Yes, that's what I meant when I said the models and config were already set up for it :) I think anything related to category pagination can be removed. There should be no impact on anyone using category pagination already unless they're using a non-default value.

pionl commented:
will check later, then I will push changes for a custom models if it's ok. Will have to refactor the changes once I change the pagination. Have a great day.
